File: android/app/src/main/java/com/dontworry/app/domain/search/SearchIndex.kt
package com.dontworry.app.domain.search

import com.dontworry.app.domain.model.Thread

data class IndexedDocument(
    val thread: Thread,
    val terms: Map<String, Int>,
    val length: Int
)

class SearchIndex private constructor(
    val documents: List<IndexedDocument>,
    private val docFrequencies: Map<String, Int>,
    val averageDocLength: Double
) {
    fun docFrequency(term: String): Int = docFrequencies[term] ?: 0

    companion object {
        fun fromThreads(threads: List<Thread>): SearchIndex {
            val documents = threads.map { thread ->
                val tokens = QueryTokenizer.tokenizeForIndex(thread.threadTitle, thread.threadContent)
                IndexedDocument(
                    thread = thread,
                    terms = tokens.groupingBy { it }.eachCount(),
                    length = tokens.size
                )
            }

            val docFrequencies = mutableMapOf<String, Int>()
            documents.forEach { doc ->
                doc.terms.keys.forEach { term ->
                    docFrequencies[term] = (docFrequencies[term] ?: 0) + 1
                }
            }

            val avgLength = if (documents.isEmpty()) 0.0 else documents.map { it.length }.average()
            return SearchIndex(documents, docFrequencies, avgLength)
        }
    }
}

File: ios/DontWorryiOS/Domain/SearchIndex.swift
import Foundation

struct IndexedDocument {
    let thread: Thread
    let terms: [String: Int]
    let length: Int
}

struct SearchIndex {
    let documents: [IndexedDocument]
    private let docFrequencies: [String: Int]
    let averageDocLength: Double

    func docFrequency(term: String) -> Int {
        docFrequencies[term] ?? 0
    }

    static func fromThreads(_ threads: [Thread]) -> SearchIndex {
        let documents = threads.map { thread in
            let tokens = QueryTokenizer.tokenizeForIndex(title: thread.threadTitle, content: thread.threadContent)
            return IndexedDocument(
                thread: thread,
                terms: Dictionary(tokens.map { ($0, 1) }, uniquingKeysWith: +),
                length: tokens.count
            )
        }

        var frequencies: [String: Int] = [:]
        for document in documents {
            for term in document.terms.keys {
                frequencies[term, default: 0] += 1
            }
        }

        let average = documents.isEmpty ? 0.0 : Double(documents.map(\.length).reduce(0, +)) / Double(documents.count)
        return SearchIndex(documents: documents, docFrequencies: frequencies, averageDocLength: average)
    }
}
